streamlit_app.py

```diff
@@ -194,3 +195,113 @@
 However flexible a BESS is, it is not this level of flexible.
 Maybe your system is charging for more than 4 hours after each other, the battery would be full far earlier than that it would be able to do that. 
 """)
+
+st.write("""
+## Version 2. Track Capacity (MWh) while steering power
+As mentioned in the previous section, batteries are flexible asset, if they would be able to magic energy out of thin air, the business case would be easy.
+However this is not the case, BESS are limited by the amount of power they have stored.
+This is called their State of Charge, or SoC for short.
+When a BESS charges, it becomes more full.
+When a BESS discharge, it becomes more empty.
+
+We add these SoC limitations to the Linear Solver as follows:
+
+    capacity = [solver.IntVar(min_battery_capacity_kwh, max_battery_capacity_kwh, f'capacity_period_{i}') for i in range(0, len(charge_power) + 1)]
+    # Capacity constraint ensure capacity follows from previous timestep
+    capacity_from_charging = charge_power[i - 1] * length_of_timestep_hour
+    capacity_from_charging = capacity_from_charging * charge_efficiency
+    capacity_from_discharging = discharge_power[i - 1] * length_of_timestep_hour
+    capacity_from_discharging = capacity_from_discharging * discharge_efficiency
+    solver.Add(capacity[i] == capacity[i - 1] + capacity_from_charging - capacity_from_discharging)
+
+A parameter `length_of_timestep_hour` has to be passed to these parameters as well.
+Looking into the formula of converting power to capacity shows us why.
+
+    capacity = power * hour
+
+This optimisation is on the dayahead market, with hourly timesteps, each power action in MW can be converted to MWh by multiplying it by 1.
+We utilise the same optimisation function, lets see how much money the model generates now:
+""")
+# ---------- VERSION 2 ALLOW POWER STEERING, TRACK CAPACITY ----------
+solver = pywraplp.Solver('VERSION 2 MAXIMIZE REVENUE, ACCOUNT FOR SoC', pywraplp.Solver.GLOP_LINEAR_PROGRAMMING)
+charge_power, discharge_power = add_power_schedules_to_solver(
+    solver=solver,
+    schedule_length=len(price_schedule_df),
+    max_power_kw=max_power_kw,
+)
+capacity, cycles = add_capacity_and_cycles_to_solver(
+    solver=solver,
+    charge_power=charge_power,
+    discharge_power=discharge_power,
+    min_battery_capacity_kwh=min_battery_capacity_kwh,
+    max_battery_capacity_kwh=max_battery_capacity_kwh,
+    initial_battery_capacity_kwh=initial_battery_capacity_kwh,
+    final_battery_capacity_kwh=final_battery_capacity_kwh,
+    length_of_timestep_hour=length_of_timestep_hour,
+    charge_efficiency=1,  # 100% RTE for now
+    discharge_efficiency=1,
```

Comment:
Would it be possible to add the parameter self-discharge? The self-discharge would only apply when the battery is not (dis)charging. This value would be a percentage discharge of the available SoC over time, for instance: 3%/month of self-discharge for li-ion batteries, during idling.

Reply:
If self-discharge acts on a month scale, I do not think it is relevant to implement ASAP into what is currently a daily optimisation model. E.g. the BESS will be so active that self-discharge (especially with li-ion) is negligble.

That being said, if this benchmark were to be updated to run easily on monthly or yearly market data, then implementing self-discharge should definitely be a consideration.